Vlan

whooo opened the pull request:
##### SUMMARY
Adds two modules to manage VLANs on SONiC.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
vlan

##### ADDITIONAL INFORMATION
Adds two module to be able to manage VLAN configuration.
The reason for splitting the functionality into two modules is that it reflects the configuration database structure (the `VLAN` and `VLAN_MEMBER` tables).
Neither module tries to merge current and wanted configuration but rather overwrites the current configuration, per table entry that is.

whooo committed:
Add vlan module

This module can manage VLANs attached to the bridge.
Changes are atomic in the sense that it will not try to merge current and wanted states
but instead overwrite current configuration.

Signed-off-by: Erik Larsson <who+github@cnackers.org>

whooo committed:
Add module to manage VLAN interface memberships

Signed-off-by: Erik Larsson <who+github@cnackers.org>

bluecmd commented on plugins/modules/vlan_member.py:
```diff
@@ -0,0 +1,136 @@
+        description: The interface to attach to the VLAN
+        required: true
+        type: str
+    tagging_mode:
```
My initial reaction to this is that this is a bool in disguise, do you see any other states than "tagged: yes" or "tagged: no"? If so, that would be an easier way to say the same thing I think.

whooo replied:
I lazily copied how it looks in the configdb :-)
How about `untagged` with false/off being the default, it would make it clearer when reading playbooks which interfaces are untagged or not (assuming most users uses tagged interfaces)?

bluecmd replied:
My problem with `untagged` is that it's a negative which tend to be harder to read (c.f. `untagged: no` vs. `tagged: yes`).

As for what the default should be, personally I would probably make it a required field to avoid confusion. I know from experience how often I forget the `-u` flag in SONiC when I want to make a port untagged - I wish they made the CLI require specifying both so you don't have to remember which is the default.

I'll leave both these choice to you since it's your PR.

whooo replied:
I went with your suggestion as I agree that making the option explicit is better

bluecmd commented on plugins/modules/vlan_member.py:
```diff
@@ -0,0 +1,136 @@
+    cur_val = config_db.get_entry('VLAN_MEMBER', key)
+    changed = compare_vlan_member_entries(val, cur_val)
+
+    if module.check_mode or not changed:
```
The behavior when the existing key contains an extra property not implemented by this module is too hidden for my taste. I would like to see a test that cements how these extra properties are handled.

Like this:
```py
VLAN_TABLE = {
    'Vlan3361': {
        'vlanid': 3361,
        'extra-option': 1234,
    }
}

_, val = vlan.build_vlan_entry(vlanid=3361, state='present')
changed = vlan.compare_vlan_entries(val, VLAN_TABLE.get('Vlan3361'))
assert not changed, 'extra options should be silently ignored'

```

What do you think?

whooo replied:
Yeah, I missed that, doing a deep comparison of both dicts is probably better

whooo replied:
I reworked it by adding a utility function for comparing two (possible empty/None) entries. so any option the modules aren't aware of will still lead to changed being true

bluecmd commented on plugins/modules/vlan.py:
```diff
@@ -0,0 +1,163 @@
+---
+module: vlan
+short_description: Configure a SONiC VLANs
+version_added: "0.1.2"
```
It will be released in version 0.2.0.

whooo replied:
Fixed

bluecmd reviewed (changes requested):
Thanks a lot for this PR! It looks great!

I found three (well, four including the version comment) things that I feel could be improved, but they are not really mandatory - we can merge with things as is if you feel strongly about it.

The first two things are inline as comments, the last one is that I would like to see an integration test as well, like here: https://github.com/ansible-collections/community.sonic/blob/main/tests/integration/targets/sonic_interface_port/tasks/main.yaml.
That way we know that the values written to Redis actually works with the rest of the system.

whooo committed:
Add module util to compare two config db entries

Signed-off-by: Erik Larsson <who+github@cnackers.org>

whooo commented:
> I found three (well, four including the version comment) things that I feel could be improved, but they are not really mandatory - we can merge with things as is if you feel strongly about it.
> 
> The first two things are inline as comments, the last one is that I would like to see an integration test as well, like here: https://github.com/ansible-collections/community.sonic/blob/main/tests/integration/targets/sonic_interface_port/tasks/main.yaml. That way we know that the values written to Redis actually works with the rest of the system.

Sorry for the push-spam, I accidentally rebased away some changes .
I added a small amount of inline comments as well as integration testing.

bluecmd commented:
Great work! Did you mean to remove the unit tests that test the `changed` logic? I liked those tests, and while we have the integration tests (great job, very nice!) they are by nature not as explicit as the unit tests are.

Let me know if you want to re-add those tests, otherwise I'll do another pass hopefully this evening and merge it then.

whooo commented:
> Great work! Did you mean to remove the unit tests that test the `changed` logic? I liked those tests, and while we have the integration tests (great job, very nice!) they are by nature not as explicit as the unit tests are.
> 
I wrote the integration tests but overwrote them by accident during a rebase so had to rewrite them :-(
I removed the change tests as I moved the functionality to an utility function and test that function instead.

> Let me know if you want to re-add those tests, otherwise I'll do another pass hopefully this evening and merge it then.
I'm good with the current state of the PR now

bluecmd commented on plugins/module_utils/entries.py:
```diff
@@ -0,0 +1,21 @@
+from ansible.module_utils.common.dict_transformations import recursive_diff
```
I think you need to add a GPL header here, according to https://docs.ansible.com/ansible/devel/community/collection_contributors/collection_requirements.html#collection-licensing-requirements

whooo replied:
Fixed

whooo commented:
I pushed a possible fix for the 2.15 sanity check error

bluecmd commented:
Thanks!
